### DISCUSSION TITLE
Significant changes mode - no location update

### DISCUSSION THREAD

**Discussion by alsparks** [owntracks/android#1071] [Ideas & New Features]
Trying out Owntracks, have set up an HTTP endpoint for receiving region monitoring.  I'm starting with Owntracks app 2.1.2 (21201) running on a Samsung S8 (Android 8.0.0).

I started out with region defined, and running in significant changes mode.   The app monitor is turned off, and I've turned off battery optimization for the app.  I tested it today, took a drive.  When I opened the phone, the notification still had my starting address, took about 15 minutes to get to destination.   I opened the app proper, and it still had my starting address on map, and did not update, even when I pushed the compass or upload buttons.

Until I changed to move move... then the location abruptly changed, and the region transition happened.   It works basically OK with move mode on - except the battery consumption is too much to bear.

Am I missing something else I need to do?  I had set the locator displacement to 150, and the location interval is still at default (900).  BTW- bug: the text on the interval option is the same as for the displacement option.

EDIT:  Also should note the notification in status bar is also enabled as required.

Thanks in advance for any advice.
-Alan

**Comment by binarybucks** [comment]
900s interval is 15 minutes so this might be expected. Try lowering the interval, removing the displacement and check if Wifi Assisted location is enabled.

**Comment by alsparks** [comment]
Tried setting the interval to 120 (assume this is in seconds), and the displacement to 0 (assume that is what you meant by "remove").   Didn't help.   Drove 15 minutes to a location well away from home, and it did not update location.  At least, not until I got there, saw the Owntracks map stuck at original location - opened Google Maps and it updated, and suddenly so did Owntracks.  Setting the interval lower obviously didn't get it to find the changing location any faster (or at all).

Checked my configuration for location - it is set as "high accuracy" with all of GPS, WiFi, and Bluetooth enabled.  Doublechecked optimization, phone set with power savings mode off and the app not battery optimized.

Is there any way to get Android location to update Owntracks in a reasonable way?  Definitely not usable in this form.

Thanks in advance, again.
-Alan

**Comment by growse** [comment]
What experience do you get running under "move mode"? That puts the location request into high power mode, so you may see an impact on battery life, but it should help give an indication as to whether the problem is with the significant changes mode, or background updates in general.

**Comment by alsparks** [comment]
I mention above that it works better, it's much more responsive and tracks better.   It just isn't usable for any appreciable length of time, due to how badly and quickly it drains the device battery.

**Comment by 4994james** [comment]
Hi.  I think this issue arose at the last update.  I am seeing the same behaviour.  It gets "stuck" on significant monitoring mode.  I also cannot see the config options for this mode any more.  This is Android version.

**Comment by 4994james** [comment]
I will try this. I am Moto g6. Android 9.

**Comment by binarybucks** [comment]
Significant changes mode defaults to "Balanced Power Mode" which uses Wifi Location as the primary location source. If other apps like Google Maps provide location data, that is reused. 
If you disable Wifi Location on your phone, Owntracks will just get information when other apps use location services. 
You can either activiate Wifi Location or set High Power Mode for signifficant changes mode in the preferences editor

**Comment by alsparks** [comment]
As the OP on the ticket... by "high power" mode I assume you mean "high accuracy" mode in location settings.  I see no "high power mode" settings in Android 8, or the app.

I have high accuracy mode set, and wifi and bluetooth scanning turned on.  But yet the spotty and slow location updates.

As a side note, my experiences to date using the 4 app modes:
* move mode - best luck with getting timely location updates.  But not usable in general due to the high battery drain.
* quiet and manual mode - docs say I should still get zone enter/leave updates... but my experiences is I do not.   I moved a lot yesterday, in and out of defined zone, with no update from the app.
* significant changes mode - where I'm at now, do get some updates, they tend to be slow even with the displacement settings changes (offset 0, interval 60).

**Comment by 4994james** [comment]
In significant changes mode the app has ceased to function as far as I can see.  Motog6 Android 9.
![Screenshot_20190501-101313](https://user-images.githubusercontent.com/44417262/57012799-2babf400-6c00-11e9-9eae-6596be3e1c92.png)

**Comment by binarybucks** [comment]
Since which app version and do you have WiFi location enabled?

**Comment by 4994james** [comment]
I know this is a bit rubbish but I am not sure.  I have updated this app recently but I have also installed an Android upgrade.  I cannot now set wifi location on or off.  It seems to be set at app level.  Maybe this is the root cause.  I will try to install an earlier version of owntracks to see if I can eliminate that

**Comment by 4994james** [comment]
Early days but I have reverted to v21201 and it worked properly this morning.  It reset all the thresholds to defaults so not sure if that is completely definitive.  I will run it for a couple of days and then try the newer version again.

**Comment by alsparks** [comment]
Where did you find the apk for that?  Interested in your findings.  
Current version is too slow and erratic for my use.

-Alan

**Comment by 4994james** [comment]
I have backups but it is also on apkpure.

**Comment by binarybucks** [comment]
https://github.com/owntracks/android/releases

**Comment by alsparks** [comment]
So I already had 212 installed, and I don't know if downgrading to less 
than that makes any sense for this.

I notice mention of a 213, but no APKs available yet for that. Don't 
know if that has any solutions, but can't test that anyway.

-Alan

**Comment by binarybucks** [comment]
@alsparks move mode uses GPS and hence works without WiFi location but uses significant battery. You can change significant mode to also use GPS. You may tweak it a bit to slightly reduce battery usage but that won’t change much.
Geofences transitions in quiet and manual mode heavily rely on WiFi location too. These modes just use a low power location request that mostly rely on Cell tower locations. As these are very inaccurate, you likely won’t get any transition events that the app calculates for every new location fix in addition to the OS geofence API. 

From my experience the Android location APIs are unable to provide accurate and timely background location updates without WiFi location. This is questionable from a privacy point of view but technically understandable because the alternate location sources are either inaccurate or use a lot of power.

**Comment by grheard** [comment]
@binarybucks I've been toying with the idea for a hybrid mode within owntracks. A mode that uses PRIORITY_BALANCED_POWER_ACCURACY by default and on a user configured time basis switches momentarily to PRIORITY_HIGH_ACCURACY for gps fix. It might bridge the gap between the loose accuracy of the balanced power mode and the high battery usage of the high accuracy mode. It will certainly use more power, but perhaps not as much as leaving high accuracy on all the time.

**Comment by binarybucks** [comment]
Good idea. You could already script this yourself using tasker and an intent to toggle the monitoring mode

**Comment by daniele-athome** [comment]
Hi, I've been having the same issue but using GPS-only location (due to privacy issue with WiFi location).
I can see that the preference `Keys.LOCATOR_PRIORITY` is not visible to the user, maybe exposing it among the other Locator preferences might be useful (in a meaningful way of course)?
I mean I'd be ok to turn on the GPS every e.g. N minutes (or when the OS decides to enter the "wake up window") in order to avoid using WiFi-based location.

**Comment by binarybucks** [comment]
There’s the preferences editor that can be used to easily change every value. 
Presently we do not plan to build a mode that periodically turns on the GPS because that can be scripted with Tasker or with remote reportLocation commands in MQTT mode.  We’re if course always accepting pull requests.

**Comment by daniele-athome** [comment]
> There’s the preferences editor that can be used to easily change every value.

I didn't notice it, thanks.
When I try to set `locatorPriority` to 3 (or 2 FWIW) it says "Invalid key specified". Also exporting and manually adding it and then importing again ignores the key. :sob: 

EDIT: S***... using version 2.1.2... sorry :angel:

**Comment by alsparks** [comment]
Alexander, as i've mentioned before, I have high accuracy mode enabled, 
with wifi location turned on.  Think your reply is confused with others 
that have posted here.  I get excessive delays in location tracking (and 
zone transition reporting, which is the only important factor for me) 
even then.

-Alan

**Comment by alsparks** [comment]
You've mentioned Tasker a couple of times, but I don't see any control 
points for Owntrack listed in Tasker.  Is this a planned feature, or 
something that exists in unreleased version?

-Alan

**Comment by binarybucks** [comment]
Indeed, I mixed it up with @pm-pm's reply. Sorry for the confusion. 
You can use Tasker to send intents to Owntracks. Theres's no dedicated plugin. 

Regarding the location issues, can you please try the 2.1.3 release from the beta channel (https://play.google.com/apps/testing/org.owntracks.android)? It fixes an issue with mode handling that could explain your problems.

**Comment by alsparks** [comment]
Thanks, I loaded and did a quick trip to try 2.1.3 relative to my 
latency issues.

Didn't really help.  It wasn't too bad this one test leaving the zone.  
But lagged badly entering the zone, was home 2-3 minutes before it 
triggered enter.  Zone radius of 250 meters.

-Alan

**Comment by timdonovanuk** [comment]
Android phones [natively provide an activity](https://developers.google.com/android/reference/com/google/android/gms/location/DetectedActivity) such as "IN_VEHICLE", "STILL", "WALKING" etc. Wouldn't it make sense to use this to drive the owntracks mode?

**Comment by growse** [comment]
Good shout - I wasn't aware of this. I'm wondering what the UX should be for this sort of thing? Perhaps a preference that (when enabled) toggles the mode from significant to move based on when the user is `STILL` or not?

I'll add it in as an experimental feature for 2.5 and do some testing.

**Comment by timdonovanuk** [comment]
> I'm wondering what the UX should be for this sort of thing? 

Hmm good question! For maximum flexibility I would think an advanced option of being able to set:

Activity | Mode | UpdateFrequency 

would be nice. 

For example that means if a user wanted, they could customise it to do "when I'm driving, I would like to be in move mode, with updates every 5 seconds" (e.g. IN_VEHICLE | MOVE | 5s), but when on foot, I only need significant updates every 5 mins ( ON_FOOT | SIGNIFICANT | 300s). But this is probably a lot of work! 

I think an auto setting based on your judgement is a great idea (I'm super new to this so have no idea what settings are best in each situation! My personal experience is I would love to have MOVE mode when in a vehicle, as currently my recorded points jump all over the place when in a car and don't show the true route). 

So yep indeed, I think an awesome first step would be toggling the mode from significant to move based on when the user is STILL or not. Thank you :)

**Comment by growse** [comment]
I hear what you're saying, I think I disagree that we should be optimizing for "maximum flexibility". Giving too many options to users means you often end up having to troubleshoot some weird behaviour, because of some configuration combination you've never tested.

> My personal experience is I would love to have MOVE mode when in a vehicle, as currently my recorded points jump all over the place when in a car and don't show the true route).

I have *exactly* this requirement, but solve it using Tasker. When it detects that the device is bluetooth connected to my car and plugged in, it sends an intent to OT to switch to move mode. When those conditions stop being true, it switches back to significant changes.

There's a general principle here about automation and orchestration, and to what extent that this is something the app should do, vs the app should allow itself to be controlled by other things. Generally, in the past, we've leaned towards the latter, because once you start supporting one type of automation, everyone's niche requirements suddenly appear (because everyone wants something slightly different) - there's already a plethora of Tasker-like apps that can do clever things based on the device context. Replicating that logic in the app itself would both add complexity and also run the risk that it actually doesn't work as well as a separate app designed exactly for this purpose. That said, I'll have a look and may end up adding it as an experimental feature - I could be convinced that this sort of functionality ("move when I'm moving, significant when I'm still") is important enough to implement.

*edit* going to play about with the tasker plugin (https://play.google.com/store/apps/details?id=io.shortway.detact) that allows driving events based on device activity.

**Comment by jsarcone** [comment]
I'm still in seeing these issues too. Samsung S10 5G Android 12.
